frontend/src/components/pages/acls/UserEditModals.tsx

```diff
@@ -0,0 +1,256 @@
+                      variant="ghost"
+                      aria-label="Refresh"
+                      icon={<MdRefresh size={16} />}
+                      display="inline-flex"
+                    />
+                  </Tooltip>
+                  <Tooltip label={'Copy password'} placement="top" hasArrow>
+                    <CopyButton content={password} variant="ghost" />
+                  </Tooltip>
+                </Flex>
+                <Checkbox
+                  isChecked={generateWithSpecialChars}
+                  onChange={(e) => {
+                    setGenerateWithSpecialChars(e.target.checked);
+                    setPassword(generatePassword(30, e.target.checked));
+                  }}
+                >
+                  Generate with special characters
+                </Checkbox>
+              </Flex>
+            </FormField>
+            <FormField label="SASL Mechanism" showRequiredIndicator>
+              <SingleSelect<SASLMechanism | undefined>
+                options={[
+                  {
+                    value: SASLMechanism.SASL_MECHANISM_SCRAM_SHA_256,
+                    label: 'SCRAM-SHA-256',
+                  },
+                  {
+                    value: SASLMechanism.SASL_MECHANISM_SCRAM_SHA_512,
+                    label: 'SCRAM-SHA-512',
+                  },
+                ]}
+                value={mechanism}
+                onChange={(e) => {
+                  setMechanism(e);
+                }}
+              />
+            </FormField>
+          </Flex>
+        </ModalBody>
+        <ModalFooter display="flex" gap={2}>
+          <Button
+            variant="ghost"
+            isDisabled={isUpdateUserPending}
+            onClick={() => {
+              setIsOpen(false);
+            }}
+          >
+            Cancel
+          </Button>
+          <Button
+            variant="solid"
+            type="submit"
+            onClick={onSavePassword}
+            isDisabled={!isValidPassword || mechanism === undefined || isUpdateUserPending}
+          >
+            Save changes
+          </Button>
+        </ModalFooter>
+      </ModalContent>
+    </Modal>
+  );
+};
+
+type ChangeRolesModalProps = {
+  userName: string;
+  isOpen: boolean;
+  setIsOpen: (isOpen: boolean) => void;
+};
+
+export const ChangeRolesModal = ({ userName, isOpen, setIsOpen }: ChangeRolesModalProps) => {
+  const toast = useToast()
+  const [selectedRoles, setSelectedRoles] = useState<string[]>([]);
+  const { mutateAsync: updateRoleMembership, isPending: isUpdateMembershipPending } = useUpdateRoleMembershipMutation();
+  const { data } = useListRolesQuery({ filter: { principal: userName } });
+  const originalRoles = data.roles.map((r) => r.name);
+
+  // biome-ignore lint/correctness/useExhaustiveDependencies: <explanation>
+  useEffect(() => {
+    setSelectedRoles([...originalRoles]);
+  }, [JSON.stringify(originalRoles)]);
```

Comment:
This is quite ugly, maybe at least provide an explanation why we need to do this, if it's documented then people can be more aware of this